gov/politeia/proposals.go

```diff
@@ -90,371 +87,509 @@ func NewProposalsDB(politeiaURL, dbPath string) (*ProposalDB, error) {
+		if err != nil {
+			return nil, fmt.Errorf("userMetadataDecode err: %v", err)
+		}
+		proposal.UserID = um.UserID
+
+		// Comments count
+		proposal.CommentsCount =
+			int32(commentsCounts[record.CensorshipRecord.Token])
+
+		// Vote summary data
+		summary := voteSummaries[proposal.Token]
+		proposal.VoteStatus = summary.Status
+		proposal.VoteResults = summary.Results
+		proposal.EligibleTickets = summary.EligibleTickets
+		proposal.StartBlockHeight = summary.StartBlockHeight
+		proposal.EndBlockHeight = summary.EndBlockHeight
+		proposal.QuorumPercentage = summary.QuorumPercentage
+		proposal.PassPercentage = summary.PassPercentage
+
+		var totalVotes uint64
+		for _, v := range summary.Results {
+			totalVotes += v.Votes
+		}
+		proposal.TotalVotes = totalVotes
+
+		// Status change metadata
+		ts, changeMsg, err := statusChangeMetadataDecode(record.Metadata)
+		if err != nil {
+			return nil, fmt.Errorf("statusChangeMetadataDecode err: %v", err)
+		}
+		proposal.PublishedAt = ts[0]
+		proposal.CensoredAt = ts[1]
+		proposal.AbandonedAt = ts[2]
+		proposal.StatusChangeMsg = changeMsg
+
+		// Append proposal after inserting the relevant data
+		proposals = append(proposals, &proposal)
 	}
 
-	defer atomic.StoreInt64(&db.lastSync, time.Now().UTC().Unix())
+	return proposals, nil
+}
 
-	// Retrieve and update all current proposals whose vote statuses is either
-	// NotAuthorized, Authorized and Started
-	numRecords, err := db.updateInProgressProposals()
+func (db *ProposalsDB) fetchTicketVoteResults(token string) (*pitypes.ProposalChartData, error) {
+	// Fetch ticket votes details to acquire vote bits options info.
+	details, err := db.client.TicketVoteDetails(ticketvotev1.Details{
+		Token: token,
+	})
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("Pi client TicketVoteDetails err: %v", err)
 	}
 
-	// Retrieve and update any new proposals created since the previous
-	// proposals were stored in the db.
-	lastProposal, err := db.lastSavedProposal()
-	if err != nil && err != storm.ErrNotFound {
-		return fmt.Errorf("lastSavedProposal failed: %v", err)
+	// Maps the vote bits option to their respective string ID.
+	voteOptsMap := make(map[uint64]string)
+	for _, opt := range details.Vote.Params.Options {
+		voteOptsMap[opt.Bit] = opt.ID
 	}
 
-	var queryParam string
-	if len(lastProposal) > 0 && lastProposal[0].TokenVal != "" {
-		queryParam = fmt.Sprintf("?before=%s", lastProposal[0].TokenVal)
-	}
-	publicProposals, err := db.fetchAPIData(queryParam)
+	tvr, err := db.client.TicketVoteResults(ticketvotev1.Results{
+		Token: token,
+	})
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("Pi client TicketVoteResults err: %v", err)
 	}
 
-	n, err := db.saveProposals(publicProposals)
-	if err != nil {
-		return err
+	// Parse reply to proposal chart data
+	var chart pitypes.ProposalChartData
+	var timestamps []int64
+	for _, v := range tvr.Votes {
+		bit, err := strconv.ParseUint(v.VoteBit, 16, 64)
```

Comment:
Docs for `CastVoteDetails` say

```go
VoteBit   string `json:"votebit"`   // Selected vote bit, hex encoded
```

**hex encoded**

Are the docs incorrect?

Reply:
hmmm yes seems like it. Will warn luke

Reply:
Do you have an example URL where I can pull the response from?

Reply:
https://proposals.decred.org/record/e2d7b7d

You can click on the magnifier that is besides the total votes, which is used to search for a specific ticket vote. It will then load a pop-up which starts to load the vote results request. then you check the response on the browser's network console logs 👍 

will be pushing some changes to address the comments soon mate

Reply:
not sure if useful but: https://github.com/decred/politeia/blob/84a52cfc0975cb1ca2df676d15e7ad0637a8ce42/politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go#L1467-L1468

Reply:
OK, so:

```
curl -X POST -d '{"token":"e2d7b7d6b837431f"}' https://proposals.decred.org/api/ticketvote/v1/results
```

gives vote records like:

```json
{
  "votes": [
    {
      "token": "e2d7b7d6b837431f",
      "ticket": "0002dcdc9f7bb52be9f5ee09549a2c4a630d844bf8c9c8e36e6b3e89ad45c56b",
      "votebit": "2",
      "address": "DsVYYG1nJKGm7ozrvh4gDnXCN1g6DexJ2KT",
      "signature": "1f76c0bb04cbdffb2bd248e79b8fbaab70228951a28d60b65bb121a76a98a90b5e11663a77ea3ff894fcfd858f9a0ff6380d70ef857370636b500c4849baeabbd2",
      "receipt": "f2f7b4a9a69882ee14b166ed1afa300c36099bbe3e31f78a70a8a811a0569d4fd46de7eacddfcec749fe778f01b97373ad58e6f2c2e824c33ca5a235bf15460f",
      "timestamp": 1626540343
    },
...
```

So the `votebit` field really is a text representation of a number, rather than just a number.  Any idea why that is?

Reply:
hmmm actually no, I'm not sure why that is.

Reply:
> I'll be adding a similar bitwise `&` with the bit options for validation like you asked above

Nice.  I see that being done in `plugins/ticketvote.voteBitVerify` with `voteDetails.Params.Options` and `voteDetails.Params.Mask`, so I guess that's the way it's supposed to be done.

https://github.com/decred/politeia/blob/84a52cfc0975cb1ca2df676d15e7ad0637a8ce42/politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go#L184

Reply:
OK so I'm guessing hex encoded means it's a hexadecimal number like "20" instead of 32.
Ah that makes sense as base 16 is passed to strconv.ParseUint.  OK so the docs looks right then.

Reply:
oh yes that's it, got confused there for a moment also 😅